test/apps/iodemo/io_demo.cc

```diff
@@ -246,46 +383,196 @@ class P2pDemoCommon : public UcxContext {
+            _pool(pool)
+        {
+            switch (memory_type) {
+            case UCS_MEMORY_TYPE_HOST:
+                _impl = new HostBufferImpl(buffer_size);
+                break;
+#ifdef HAVE_CUDA
+            case UCS_MEMORY_TYPE_CUDA:
+                _impl = new CudaBufferImpl(buffer_size);
+                break;
+            case UCS_MEMORY_TYPE_CUDA_MANAGED:
+                _impl = new CudaManagedBufferImpl(buffer_size);
+                break;
+#endif
+            default:
+                LOG << "ERROR: Unsupported memory type requested: "
+                    << memory_type_to_string(memory_type);
+                abort();
+            }
+
+            _impl->init();
+        }
+
+        void *buffer()
+        {
+            return _impl->buffer();
+        }
+
+        size_t capacity() const
+        {
+            return _impl->capacity();
+        }
+
+        size_t size() const
+        {
+            return _impl->size();
+        }
+
+        inline void resize(size_t size)
+        {
+            _impl->resize(size);
+        }
+
+        void release()
+        {
+            _pool.put(this);
+        }
+
+
+        ~Buffer()
+        {
+            delete _impl;
+        }
 
     private:
-        void*                     _buffer;
-        size_t                    _size;
-        MemoryPool<Buffer, true>& _pool;
+        BufferMemoryPool<Buffer> &_pool;
+        BufferImpl *_impl;
+    };
+
+
+    class HostBufferImpl : public BufferImpl {
+    public:
+        HostBufferImpl(size_t size) : BufferImpl(size)
+        {
+        }
+
+        virtual void init()
+        {
+            setBuffer(memalign(ALIGNMENT, _capacity));
+        }
+
+        ~HostBufferImpl()
+        {
+            free(buffer());
+        }
     };
 
+#ifdef HAVE_CUDA
+    class CudaMemoryBufferImpl : public BufferImpl {
+    public:
+        CudaMemoryBufferImpl(size_t size) : BufferImpl(size)
+        {
+        }
+
+        virtual void init() override
+        {
+            void *tmp;
+            cudaError_t cerr = allocate(&tmp, _capacity);
+            if (cerr != cudaSuccess) {
+                tmp = NULL;
+            }
+            setBuffer(tmp);
+        }
+
+        virtual cudaError_t allocate(void **ptr, size_t size) = 0;
+        void free()
+        {
+            cudaFree(buffer());
+        }
+
+        ~CudaMemoryBufferImpl()
+        {
+        }
+    };
+
+    class CudaBufferImpl : public CudaMemoryBufferImpl {
+        using CudaMemoryBufferImpl::CudaMemoryBufferImpl;
+
+    public:
+        ~CudaBufferImpl()
+        {
+            free();
+        }
+
+        virtual cudaError_t allocate(void **ptr, size_t size)
+        {
+            return cudaMalloc(ptr, size);
+        }
+    };
+
+    class CudaManagedBufferImpl : public CudaMemoryBufferImpl {
+        using CudaMemoryBufferImpl::CudaMemoryBufferImpl;
+
+    public:
+        ~CudaManagedBufferImpl()
+        {
+            free();
+        }
```

Comment:
this buffer thing is getting too complicated 
can't we just have one buffer class that gets the mem type in its constructor?

Reply:
That was the initial implementation, then after @evgeny-leksikov review I made the classes:
https://github.com/openucx/ucx/pull/6171#discussion_r563687760

From design perspective, perhaps there is something good in having these classes which implement the same interface in different ways.

Shall I revert and remove the classes?